### PR TITLE
Add Alpha Dropout

### PR DESCRIPTION
**Pull request by fmassa**
Complements https://github.com/pytorch/pytorch/issues/1768.

Note that I have not added the `fixed_mean` and `fixed_var` parameters from the original paper, as SELU outputs have zero mean and unit std, so I only considered this case. Let me know if you think it's worth adding support for these extra parameters.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -406,6 +407,28 @@ def dropout(input, p=0.5, training=False, inplace=False):
...
+                         "but got {}".format(p))
+
+    if p > 0 and training:
+        alpha = -1.7580993408473766
+        keep_prob = 1 - p
+        noise = input.data.new(input.size())
+        noise.bernoulli_(keep_prob)
+        noise = Variable(noise)
+
+        output = (input * noise).add_(noise.neg().add_(1).mul_(alpha))
```
Comment (fmassa) [inline torch/nn/functional.py]: is it worth compromising readability in this case for smaller memory footprint? I suppose with jit compilation these point-wise expressions will be compiled to a single kernel anyway.

File: torch/nn/functional.py [hunk 2]
```diff
@@ -406,6 +407,28 @@ def dropout(input, p=0.5, training=False, inplace=False):
     return _functions.dropout.Dropout(p, training, inplace)(input)
 
 
+def alpha_dropout(input, p=0.5, training=False):
+    if not 0 < p <= 1:
+        raise ValueError("dropout probability has to be between 0 and 1, "
+                         "but got {}".format(p))
+
+    if p > 0 and training:
```
Comment (apaszke) [inline torch/nn/functional.py]: wouldn't it be simpler to do (no need to clone the input btw)
```python
if p == 0 or not training:
    return input

# complicated case here
```

File: torch/nn/functional.py (hunk 1 above)
Comment (apaszke) [inline torch/nn/functional.py]: How about
```python
output = input * noise
torch.eq(noise, 0, out=noise) # invert noise
output.add_(alpha, noise)
```

File: torch/nn/modules/dropout.py [hunk 3]
```diff
@@ -153,3 +153,54 @@ def __repr__(self):
...
+        >>> m = nn.AlphaDropout(p=0.2)
+        >>> input = autograd.Variable(torch.randn(20, 16))
+        >>> output = m(input)
+
+    .. _Self-Normalizing Neural Networks: https://arxiv.org/abs/1706.02515
+    """
+
+    def __init__(self, p=0.5):
+        super(AlphaDropout, self).__init__()
+        if p < 0 or p > 1:
```
Comment (apaszke) [inline torch/nn/modules/dropout.py]: These checks should be consistent between module and a function.

File: test/test_nn.py [hunk 4]
```diff
@@ -763,6 +763,23 @@ def test_Dropout3d(self):
         input = torch.Tensor(num_features, b, d, w, h)
         self._test_dropout(nn.Dropout3d, input)
 
+    def test_AlphaDropout(self):
+        # generate random tensor with zero mean and unit std
+        input = torch.randn(5000)
+
+        mean = input.mean()
+        std = input.std()
```
Comment (apaszke) [inline test/test_nn.py]: Shouldn't we just use 0 and 1 for these?

File: torch/nn/functional.py (hunk 1 above)
Comment (fmassa) [inline torch/nn/functional.py]: Not sure we can add an output argument to `Variables`, can we?

File: test/test_nn.py (hunk 4 above)
Comment (fmassa) [inline test/test_nn.py]: Yes, but I left it like that in the event we might want to add support for a more general Alpha Dropout, which accepts a `mean` and `std` parameter. Also, because the samples are not infinitely large, I'd eventually need to increase the tolerance (or the tensor size) for the checks to consistently pass if I set those values to 0 and 1.

File: torch/nn/functional.py (hunk 1 above)
Comment (fmassa) [inline torch/nn/functional.py]: Also, `noise` is required for backward, I can't do an operation in-place in it (and this looks in-place, right?).

Comment (fmassa) [inline torch/nn/functional.py]: BTW, I just noticed that we don't have the overload of `add_` corresponding to `x + alpha * y` for Variables. Is this something we would want to add?

File: torch/nn/functional.py [hunk 5]
```diff
@@ -406,6 +407,28 @@ def dropout(input, p=0.5, training=False, inplace=False):
...
+    if p == 0 or not training:
+        return input
+
+    alpha = -1.7580993408473766
+    keep_prob = 1 - p
+    noise = input.data.new(input.size())
+    noise.bernoulli_(keep_prob)
+    noise = Variable(noise)
+
+    output = (input * noise).add_(noise.neg().add_(1).mul_(alpha))
```
Comment (apaszke) [inline torch/nn/functional.py]: Actually, isn't this the same as:
```python
...
noise.bernoulli_(drop_prob) # NOTE: drop_prob, not keep_prob
...
output = input.masked_fill(noise, alpha)
```

Comment (fmassa) [inline torch/nn/functional.py]: Yes, this is much better!

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks Francisco!

File: torch/nn/functional.py [hunk 6]
```diff
@@ -406,6 +407,28 @@ def dropout(input, p=0.5, training=False, inplace=False):
...
 
+def alpha_dropout(input, p=0.5, training=False):
+    if p < 0 or p > 1:
+        raise ValueError("dropout probability has to be between 0 and 1, "
+                         "but got {}".format(p))
+
+    if p == 0 or not training:
+        return input
+
+    alpha = -1.7580993408473766
```
Comment (andreh7) [inline torch/nn/functional.py]: this looks like alpha prime = - lambda * alpha from the paper (arxiv:1706.02515). Why is this a fixed value (in the paper it looks like it depends on the preceding SELU activation layer) and what expression leads to this numerical value ?

Comment (fmassa) [inline torch/nn/functional.py]: I've used the value from [original implementation by the authors](https://github.com/bioinf-jku/SNNs/blob/master/selu.py#L35), which seems to be the value so that the distributions are with mean 0 and std 1 after each layer. I could add arguments to enable the user to specify the desired mean/std, but is there an use-case for that?

Comment (andreh7) [inline torch/nn/functional.py]: Thanks for the link. For consistency, wouldn't it be better to  set `alpha = - SELU.alpha * SELU.scale' (I see in #1769 that the corresponding values are fixed in class `SELU`) ? The day somebody generalizes class `SELU` at least they will notice that `alpha_dropout()` needs to be adapted.
